File: partition-array-for-maximum-sum.cc
// Partition Array for Maximum Sum
#define REP1(i, n) for (remove_cv<remove_reference<decltype(n)>::type>::type i = 1; i <= (n); i++)
#define ROF(i, a, b) for (remove_cv<remove_reference<decltype(b)>::type>::type i = (b); --i >= (a); )

class Solution {
public:
  int maxSumAfterPartitioning(vector<int>& A, int K) {
    int n = A.size();
    vector<int> s(n+1);
    REP1(i, n) {
      int mx = 0;
      ROF(j, max(0, i-K), i) {
        mx = max(mx, A[j]);
        s[i] = max(s[i], s[j]+(i-j)*mx);
      }
    }
    return s[n];
  }
};

File: partition-array-for-maximum-sum.rs
impl Solution {
    pub fn max_sum_after_partitioning(arr: Vec<i32>, k: i32) -> i32 {
        let n = arr.len();
        let mut dp = vec![0; n+1];
        for i in 0..n {
            let mut mx = 0;
            for j in ((if i < k as usize {0} else {i+1-k as usize})..=i).rev() {
                mx = mx.max(arr[j]);
                dp[i+1] = dp[i+1].max(dp[j]+mx*(i-j+1) as i32);
            }
        }
        dp[n]
    }
}
